runtime/gc_structs/ClassArrayClassSlotIterator.cpp

```diff
@@ -46,10 +46,27 @@ GC_ClassArrayClassSlotIterator::nextSlot()
 			classPtr = (J9Class *)_iterateClazz->arrayClass;
 			if (!_isArrayClass) {
 				_state = classArrayClassSlotIterator_state_done;
+				break;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			} else if (NULL == classPtr) {
+				classPtr = (J9Class *)_iterateClazz->nullRestrictedArrayClass;
+				_state += classArrayClassSlotIterator_state_componentType;
```

Comment:
`_state` should not be incremented by `classArrayClassSlotIterator_state_componentType`.

Reply:
It is possible that `classPtr` (`_iterateClazz->nullRestrictedArrayClass`) is null and it is returned incorrectly from here.

Reply:
This iterator is written in style, where if class pointer to return is NULL we are going to the next state. Do we need to restrict to report `nullRestrictedArrayClass` otherwise `companionArray`. Can (and should) we read both unconditionally and probably rely on one or another is NULL? If both of this fields can be non-NULL should we report both to have proper class liveness discovery?

Reply:
If I understand intention correctly, I think the only change in this method we should add two extra states to return class pointers from `nullRestrictedArrayClass` and `companionArray` fields, no extra logic required. State `classArrayClassSlotIterator_state_nullRestrictedArrayClass` and handling for it is proposed already.